facefusion/face_analyser.py

```diff
@@ -164,64 +164,42 @@ def detect_with_retinaface(temp_frame : Frame, temp_frame_height : int, temp_fra
+	temp_frame = numpy.ascontiguousarray(temp_frame[..., ::-1].transpose(2, 0, 1)[None].astype(numpy.float32) / 255.0)
 	with THREAD_SEMAPHORE:
-		pred = face_detector.run(None, {face_detector.get_inputs()[0].name: temp_frame})
-	pred = numpy.squeeze(pred)
-	pred = numpy.transpose(pred, (1, 0))
-	pred = numpy.ascontiguousarray(pred)
-	bbox, score, kps = numpy.split(pred, [4, 5], axis=1)
-	indices_above_threshold = numpy.where(score > facefusion.globals.face_detector_score)[0]
-	bbox = bbox[indices_above_threshold]
-	score = score[indices_above_threshold]
-	kps = kps[indices_above_threshold]
-	new_ratio = 1/ratio
-	x_center, y_center, width, height = bbox[:, 0], bbox[:, 1], bbox[:, 2], bbox[:, 3]
-	x_min = (x_center - (width / 2) - dw) * new_ratio
-	y_min = (y_center - (height / 2) - dh) * new_ratio
-	x_max = (x_center + (width / 2) - dw) * new_ratio
-	y_max = (y_center + (height / 2) - dh) * new_ratio
-	bbox = numpy.stack((x_min, y_min, x_max, y_max), axis=1)
-	for i in range(kps.shape[1] // 3):
-		kps[:, i * 3] = (kps[:, i * 3] - dw) * new_ratio
-		kps[:, i * 3 + 1] = (kps[:, i * 3 + 1] - dh) * new_ratio
-	bbox_list = []
-	for box in bbox:
-		bbox_list.append(numpy.array(
-		[
-			box[0],
-			box[1],
-			box[2],
-			box[3],
-		]))
-	score_list = score.ravel().tolist()
-	kps_list = []
-	for keypoints in kps:
-		kps_xy = []
-		for i in range(0, len(keypoints), 3):
-			kps_xy.append([keypoints[i], keypoints[i+1]])
-		kps_list.append(numpy.array(kps_xy))
-	for bbox, keypoints in zip(bbox_list, kps_list):
-		start_point = (int(bbox[0]), int(bbox[1]))
-		end_point = (int(bbox[2]), int(bbox[3]))
-		temp_frame = cv2.rectangle(temp_frame, start_point, end_point, (0, 255, 0), 2)
-		for kp in keypoints:
-			x, y = int(kp[0]), int(kp[1])
-			temp_frame = cv2.circle(temp_frame, (x, y), 2, (0, 0, 255), -1)
+		detections = face_detector.run(None,
+		{
+			face_detector.get_inputs()[0].name: temp_frame
+		})
+	detections = numpy.ascontiguousarray(numpy.squeeze(detections).T)
+	bbox_raw, score_raw, kps_raw = numpy.split(detections, [4, 5], axis=1)
+	keep_indices = numpy.where(score_raw > facefusion.globals.face_detector_score)[0]
+	if keep_indices.any():
+		bbox_raw, kps_raw, score_raw = bbox_raw[keep_indices], kps_raw[keep_indices], score_raw[keep_indices]
+		half_width = bbox_raw[:, 2] / 2
+		half_height = bbox_raw[:, 3] / 2
+		bbox_list = [numpy.array(
+			[
+				(bbox_raw[i, 0] - half_width[i] - dw) / ratio,
+				(bbox_raw[i, 1] - half_height[i] - dh) / ratio,
+				(bbox_raw[i, 0] + half_width[i] - dw) / ratio,
+				(bbox_raw[i, 1] + half_height[i] - dh) / ratio
+			])
+			for i in range(len(bbox_raw))]
+		kps_raw[:, 0::3] = (kps_raw[:, 0::3] - dw) / ratio
+		kps_raw[:, 1::3] = (kps_raw[:, 1::3] - dh) / ratio 
+		kps_list = [numpy.array([[kps[j], kps[j+1]] for j in range(0, len(kps), 3)]) for kps in kps_raw]
```

Comment:
readability wins over code reduction - multiple for in one line are very hard to understand